### PR TITLE
Hacky friday DLCM metrics

### PR DESCRIPTION
**Pull request by pato23arg**
2x Metrics added to DLCM plugin::
***********************************
a) nautobot_lcm_devices_eos_per_part_number - end of support devices | part_id
b) nautobot_lcm_devices_eos_per_site - end of support devices | site

```
# TYPE nautobot_lcm_devices_eos_per_part_number gauge
nautobot_lcm_devices_eos_per_part_number{part_number="WS-SUP720-3BXL"} 38.0
nautobot_lcm_devices_eos_per_part_number{part_number="dcs-7048-t"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="DCS-7150S-24"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="DCS-7280CR2-60"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="veos"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="CSR1000V"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="WS-C6509-E"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="N9K-C9372TX"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="cisco-nx-osv-chassis"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="vmx"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="ccr1009-7g-1c-1splus"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="ccr1009-7g-1c-1spluspc"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="RB1200"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="ICX7150-24-2x10G"} 0.0
nautobot_lcm_devices_eos_per_part_number{part_number="vSZ6.1.1"} 0.0
# HELP nautobot_lcm_devices_eos_per_site Nautobot LCM Devices EOS per Site
# TYPE nautobot_lcm_devices_eos_per_site gauge
nautobot_lcm_devices_eos_per_site{site="ams01"} 1.0
nautobot_lcm_devices_eos_per_site{site="ang01"} 1.0
nautobot_lcm_devices_eos_per_site{site="atl01"} 1.0
nautobot_lcm_devices_eos_per_site{site="atl02"} 1.0
nautobot_lcm_devices_eos_per_site{site="azd01"} 1.0
nautobot_lcm_devices_eos_per_site{site="bkk01"} 1.0
nautobot_lcm_devices_eos_per_site{site="bre01"} 1.0
nautobot_lcm_devices_eos_per_site{site="can01"} 1.0
nautobot_lcm_devices_eos_per_site{site="cdg01"} 1.0
nautobot_lcm_devices_eos_per_site{site="cdg02"} 1.0
nautobot_lcm_devices_eos_per_site{site="del01"} 1.0
nautobot_lcm_devices_eos_per_site{site="den01"} 1.0
nautobot_lcm_devices_eos_per_site{site="dfw"} 0.0
nautobot_lcm_devices_eos_per_site{site="dfw01"} 1.0
nautobot_lcm_devices_eos_per_site{site="dfw02"} 1.0
nautobot_lcm_devices_eos_per_site{site="dxb01"} 1.0
nautobot_lcm_devices_eos_per_site{site="dxb02"} 1.0
nautobot_lcm_devices_eos_per_site{site="fra01"} 1.0
nautobot_lcm_devices_eos_per_site{site="hkg01"} 1.0
nautobot_lcm_devices_eos_per_site{site="hnd01"} 1.0
nautobot_lcm_devices_eos_per_site{site="hnd02"} 1.0
nautobot_lcm_devices_eos_per_site{site="hou"} 0.0
nautobot_lcm_devices_eos_per_site{site="icn01"} 1.0
nautobot_lcm_devices_eos_per_site{site="jcy"} 1.0
nautobot_lcm_devices_eos_per_site{site="jfk01"} 1.0
nautobot_lcm_devices_eos_per_site{site="lax01"} 1.0
nautobot_lcm_devices_eos_per_site{site="lax02"} 1.0
nautobot_lcm_devices_eos_per_site{site="lhr01"} 1.0
nautobot_lcm_devices_eos_per_site{site="lhr02"} 1.0
nautobot_lcm_devices_eos_per_site{site="nyc"} 1.0
nautobot_lcm_devices_eos_per_site{site="nym01"} 1.0
nautobot_lcm_devices_eos_per_site{site="ord01"} 1.0
nautobot_lcm_devices_eos_per_site{site="ord02"} 1.0
nautobot_lcm_devices_eos_per_site{site="pek01"} 1.0
nautobot_lcm_devices_eos_per_site{site="pek02"} 1.0
nautobot_lcm_devices_eos_per_site{site="pvg01"} 1.0
nautobot_lcm_devices_eos_per_site{site="pvg02"} 1.0
nautobot_lcm_devices_eos_per_site{site="sin01"} 1.0
nautobot_lcm_devices_eos_per_site{site="slc01"} 1.0
nautobot_lcm_devices_eos_per_site{site="weehawken"} 1.0
```

adds to #143

### CODE REVIEW

**Comment by Kircheneer** [comment]
We will have to see how this interacts with this PR: https://github.com/nautobot/nautobot-plugin-device-lifecycle-mgmt/pull/148

**Comment by pato23arg** [comment]
> We will have to see how this interacts with this PR: #148

I guess we'll have to manually update the list: metrics = [nautobot_metrics_dlcm_eos] to include all the function metrics. Do you foresee any other change when the merge occurs?

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (nkallergis) [inline nautobot_device_lifecycle_mgmt/metrics.py:14]:
```suggestion
    """Calculate number of End-of-Support (EOS) devices per Device Type and per Site.
```
I think it'll be clearer to those not familiar with the term (which is anyway a bit ambiguous as it could also mean End-of-Sales).

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (nkallergis) [inline nautobot_device_lifecycle_mgmt/metrics.py:32]:
```suggestion
            part_number = notice.device_type.part_number if notice.device_type.part_number else notice.device_type.slug
```
What if a Device Type has no Part Number entered (which is optional attribute)?

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (nkallergis) [inline nautobot_device_lifecycle_mgmt/metrics.py:42]:
```suggestion
            eos_devices_count = Device.objects.filter(device_type=notice.device_type).count()
        elif notice.inventory_item:
            part_number = notice.inventory_item
            eos_devices_count = InventoryItem.objects.filter(part_id=notice.inventory_item).count()
        else:
            continue

        part_number_gauge.add_metric(labels=[part_number], value=eos_devices_count)
```

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (nkallergis) [inline nautobot_device_lifecycle_mgmt/metrics.py:44]: Since the `yield` is outside of the for loop, what will happen if there are no HW notices?

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (nkallergis) [inline nautobot_device_lifecycle_mgmt/metrics.py:54]: Same question as above :)

**Comment by nkallergis** [comment]
Couple of minor notes here and there but very good job otherwise!

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (pato23arg) [reply nautobot_device_lifecycle_mgmt/metrics.py:42]: @nkallergis what is the goal of this item?

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (pato23arg) [reply nautobot_device_lifecycle_mgmt/metrics.py:32]: true, SKUs are important but not everybody adds them to the SoT. + to commit!

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (pato23arg) [reply nautobot_device_lifecycle_mgmt/metrics.py:44]: Correct, the way I thought of these 2 metrics was to answer 2 typical net arch Questions when facing a refresh/netop initiative:
a) how many devices per site are out of vendor support. -> spot most compromised sites
b) what type of devices are end of support -> spot most compromised platforms

do you suggest adding a third metric that shows eos platforms unfiltered? Once a platform is EOS :
EOS devices for that platform == Platform.device.count()

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (nkallergis) [reply nautobot_device_lifecycle_mgmt/metrics.py:42]: Sorry for this, bad comment on my side on multiple fronts! I intended to
1.  Remove the `metric_value = eos_devices.count()` lines by moving `.count()` to the previous ones. Not because one approach is _better_ than the other, but because you're using one-liners in lines 47-50 and it would be nice to be consistent.
1. Remove the `else-continue` clause - this shouldn't be required as either a Device Type or an Inventory Item is mandatory for a HW Notice, right?

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (nkallergis) [reply nautobot_device_lifecycle_mgmt/metrics.py:44]: Actually no: I agree with the two questions that you're trying to answer.

That said (and if I'm not forgetting something), if a Site has _no_ EoS devices/inventory items, then it yields a `nautobot_lcm_devices_eos_per_site` metric line with a value of `0.0`. Something like:
```
nautobot_lcm_devices_eos_per_site{site="ang01"} 0.0
```

This is not the case with the `nautobot_lcm_devices_eos_per_part_number` metric: if there's no device with the specific part number, there's no yielded metric line. I would expect something like:
```
nautobot_lcm_devices_eos_per_part_number{part_number="DCS-7150S-24"} 0.0
```

Let me know if that's clear and if/how I could further assist you with this.

**Comment by jvanderaa** [comment]
@pato23arg how are we looking on this?

**Comment by pato23arg** [comment]
> @pato23arg how are we looking on this?
last commit should resolve all open reviews!

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (nkallergis) [inline nautobot_device_lifecycle_mgmt/metrics.py:54]: Nice, that's the goal. Regarding the specific implementation, I think that it would be simpler and easier to understand if you limited the for loops by starting with Device Types instead of HW Notices. Something like the (untested) code below could be helpful.
```
for dt in DeviceType.objects.all():
    metric_value = Device.objects.filter(device_type=dt).count() if dt in hw_eos_device_types else 0
```

Let me know if you'd prefer to have a joint session!

File: nautobot_device_lifecycle_mgmt/metrics.py
Comment (progala) [inline nautobot_device_lifecycle_mgmt/metrics.py:12]:
```suggestion
def metrics_lcm_hw_end_of_support():  # pylint: disable=too-many-locals
```

**Review by progala — CHANGES_REQUESTED** [review]

**Review by progala — APPROVED** [review]

**Comment by jedelman8** [comment]
Is this ready to merge after being approved by @progala ?

**Comment by pato23arg** [comment]
> Is this ready to merge after being approved by @progala ?

Only docs updates are pending and then will be ready to be merged!

**Review by bradh11 — APPROVED** [review]

**Pull request merged**
